[SYCL][Doc] Add ESIMD design and implementation notes (#7824)
intel/llvm: sycl/doc/design/ESIMDDesignNotes.md

aspects, according to the "optional device features" [design](https://github.com/intel/llvm/blob/sycl/sycl/doc/design/OptionalDeviceFeatures.md).
This might require splitting implementations into per-architecture variants.
`if_device_has` [feature](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_device_if.asciidoc) may help avoid
duplication of common parts and dispatch to architecture-dependent code at
fine-grained level from within a function.
1. As VC BE moves away from `genx.*` intrinsics replacing them with
`__spirv_*` ones defined in various extensions, ESIMD should catch up.
1. Unification of common `simd_view`/`simd` interfaces in fact leads to
signigicant complication of implementation rather than its intended
simplification via avoiding code duplication, might make sense to have separate
implementations.

## Directions

This section lists possible directions for ESIMD improvements.

1. Support `std::simd`. This is the standard C++ way for explicit SIMD programming. Can help run (subsest of ESIMD) on CPU efficiently in the future.
1. Clear (via namespace?) separation of ESIMD APIs into portable and
architecture-specific parts.

> 1. Standardizing `simd_view` or equivalent. This is effectively a reference to a *subset* of `esimd::simd` vector object�s elements. The subset is defined in a regular way via starting offset, stride and number of elements in the subset. This proved to be very useful and loved by users. Missing in `std::simd`.

Author: ```suggestion
1. Standardizing `simd_view` or equivalent. This is effectively a reference
    to a *subset* of `esimd::simd` vector object's elements. The subset is defined
    in a regular way via starting offset, stride and number of elements in the subset.
    This proved to be very useful and loved by users. Missing in `std::simd`.
```

This documents is a collection of notes describing design and/or implementation
of various parts of the ESIMD programming model support within the DPC++.

## Overview of ESIMD support in DPC++ components.
ESIMD support is spread across a number of components in the oneAPI software stack, spanning compile time, link time and runtime. The picture below shows
simplified view of the DPC++ compiler and runtime diagram and where ESIMD
(sub-)components fit in it.

![Esimd Components](images/esimd_components.png)

### User program
User program can contain both SYCL and ESIMD kernels, either in the same or
different translation units. DPC++ ESIMD support will automatically split the
device code into SYCL and ESIMD parts to redirect them to different back-ends.
To facilitate this splitting, compiler will automatically identify markup and
clone parts of the ESIMD callgraph starting from kernels and functions
explicitly marked with the `intel::sycl_explicit_simd` attribute.

### Device headers
There are two distinct parts of ESIMD support in the SYCL device headers. The
first one is ESIMD-related "tweaks" within the usual SYCL headers, and the
second is the ESIMD APIs themselves.

> ##### ESIMD tweaks in SYCL device headers

Author: ```suggestion
#### ESIMD tweaks in SYCL device headers
```

The most
important one is device-side definition of the sycl::accessor class definition.
It has different layout for SYCL and ESIMD (but the same size due to padding),
as well as differnt initialization function used to assemble accessor object on
the device side from incoming kernel arguments. For SYCL, the function name is
`sycl::accessor::__init`, for ESIMD it is `sycl::accessor::__init_esimd`.
The difference is caused by the ESIMD limitation - it does not support offset,
memory and access range for an accessor object.

There are also couple ESIMD-specific proxy classes used by ESIMD API 
implementations to access internals of ESIMD objects:
- `sycl::ext::intel::esimd::detail::AccessorPrivateProxy`
this one is used by ESIMD memory APIs to access internals of `sycl::accessor`
objects
- `sycl::ext::intel::esimd::detail::WrapperElementTypeProxy`
used to access internals of SYCL types such as
`sycl::detail::half_impl::half`

> ##### ESIMD API headers

Author: ```suggestion
#### ESIMD API headers
```

compiler to obtain compile-time details for the kernel via inclusion of the
integration header generated by the generator. For ESIMD kernels `isESIMD()`
returns true, and this is used in SYCL RT headers to distinguish ESIMD kernels
from SYCL kernels, e.g. to customize kernel argument setting for accessors.
- defines `sycl_esimd_vectorize` attribute propagated to the VE BE, aimed at
 making it possible to call SIMT function from ESIMD code while vectorizing the
 former. Used internally to implement certain math functions in SYCL spec
 conformant manner (proper precision) via vectorizing scalar conformant
 implementations - see example
 [usage](https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/ext/intel/esimd/detail/math_intrin.hpp#L199)

Source locations:<br>
<tt>clang/lib/Sema/SemaSYCL.cpp</tt><br>
<tt>clang/lib/Sema/SemaDecl.cpp</tt><br>
<tt>clang/lib/Sema/SemaExpr.cpp</tt><br>
<tt>clang/include/clang/Basic/Attr.td</tt><br>
<tt>clang/include/clang/Basic/DiagnosticSemaKinds.td</tt><br>

#### Code (LLVMIR) generator

ESIMD-specific code generator tweaks are mostly translations of internal FE 
representation of variaous ESIMD attributes into LLVM IR attributes or metadata.

> ###### Kernel signature generation

Author: ```suggestion
##### Kernel signature generation
```

For ESIMD kernes, a number of additional attributes are generated for the kernel
function itself as well as certain argument.
- Kernels are annotated with `sycl_explicit_simd` and
`intel_reqd_sub_group_size` attributes. The latter must always be `1` for
a ESIMD kernel or function.
- An argument which conveys accessor's pointer is assigned a `kernel_arg_accessor_ptr` attribute

> ###### Global variable code generation.

Author: ```suggestion
##### Global variable code generation.
```

ESIMD supports "private globals" - global variables which have one copy per
thread of execution (similar to C++ thread_local), normally allocated of Gen
register file. To make a global variable a "private global", ` __attribute__((opencl_private)) __attribute__((sycl_explicit_simd))` attributes are
used. Globals of this can be forced to a specific register using the
`__attribute__((register_num(n)))` attribute. The clang code generator
translates these to `genx_volatile` and `genx_byte_offset` LLVM IR
attributes.

> ###### Function attributes translations:

Author: ```suggestion
##### Function attributes translations:
```

- Defines `intel::sycl_explicit_simd` function and variable attribute to markup
  kernels top-level ESIMD external functions and "private globals".
- Performs various ESIMD-specific semantic analysis and diagnostics. All kinds
  of ESIMD error messages can be found in
  `clang/include/clang/Basic/DiagnosticSemaKinds.td`.
- Integration header generator inserts additional
  `static constexpr bool isESIMD()` function into kernel discriptor type, which
  is used by the host compiler to obtain compile-time details for the kernel via
  inclusion of the integration header generated by the generator. For ESIMD
  kernels `isESIMD()` returns true, and this is used in SYCL RT headers to
  distinguish ESIMD kernels from SYCL kernels, e.g. to customize kernel argument
  setting for accessors.
- defines `sycl_esimd_vectorize` attribute propagated to the VC BE, aimed at
  making it possible to call SIMT function from ESIMD code while vectorizing the
  former. Used internally to implement certain math functions in SYCL spec
  conformant manner (proper precision) via vectorizing scalar conformant
  implementations - see example

>   [usage](https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/ext/intel/esimd/detail/math_intrin.hpp#L199)

Reviewer: I think we should move this type of documentation the source code and use Doxygen to produce rendered version. It's much more convenient to track the correlation between the code and the documentation. The links and source location pointers are easy to invalidate with code changes.

#### ESIMD API restriction verifier

This component is an LLVM IR pass over a compiled translation unit. It checks
for presence of certain SYCL APIs which are disallowed within ESIMD code. For
exaple, SYCL reductions are not allowed in ESIMD. The verifier does this by
demangling all the call targets within ESIMD code and matching them with
internal sub-string filters. Invoked from `clang/lib/CodeGen/BackendUtil.cpp`.

Source locations:

- `llvm/lib/SYCLLowerIR/ESIMD/ESIMDVerifier.cpp`

### SYCL post-link tool

As a part of the input device code module transformation pipeline, the
post-link tool splits the input module (or modules resulting from splitting
by other characteristics, such as aspects) into two - SYCL and ESIMD ones.
Shared functions invoked both from SYCL and ESIMD are cloned during the process.
This is necessary because SYCL and ESIMD parts must undergo different set of
transformations before generating resulting SPIR-V. ESIMD modules resulting from
splitting are marked with specific device binary property `isEsimdImage` (see
[source](https://github.com/intel/llvm/blob/9dc14a292f479880b5ab2e95f99a8414b31d1165/llvm/tools/sycl-post-link/sycl-post-link.cpp#L453)
.)

> #### Driver

Reviewer: This section should be renamed. We already have another component with the name driver and any other document references `sycl-post-link` as a driver. Probably we can merge this sub-section the parent section above.
@v-klochkov, what do you think?

Reviewer: Done in https://github.com/intel/llvm/pull/7824/commits/1b3a19f059878edb03ce1314b28dbd4d9dfc1871. Please, let me know if it looks good to you.

Reviewer 2: Looks good to me. Thank you for doing these additional changes!
